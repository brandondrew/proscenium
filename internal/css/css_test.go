package css_test

import (
	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
)

var _ = Describe("Internal/Css/Css", func() {
	Describe("ParseCss", func() {

		It("should pass through regular css", func() {
			Expect("body{}").To(BeParsedTo("body{}", "/foo.css"))
		})

		Describe("mixins", func() {
			Describe("local", func() {
				It("undefined mixin is passed through", func() {
					Expect(`
						header {
							@mixin foo;
						}
					`).To(BeParsedTo(`
						header {
							@mixin foo;
						}
					`), "/foo.css")
				})

				It("mixin not defined at root level is passed through", func() {
					Expect(`
						header {
							@define-mixin large-button {
								color: red;
							}
							div {
								@mixin foo;
							}
						}
					`).To(BeParsedTo(`
						header {
							@define-mixin large-button {
								color: red;
							}
							div {
								@mixin foo;
							}
						}
					`, "/foo.css"))
				})

				It("mixin is replaced with defined mixin", func() {
					Expect(`
						@define-mixin large-button {
							font-size: 20px;
							div { color: pink; }
						}
						header {
							@mixin large-button;
						}
					`).To(BeParsedTo(`
						header {
							font-size: 20px;
							div { color: pink; }
						}
					`, "/foo.css"))
				})

				It("dependencies are fully parsed", func() {
					Expect(`
						@define-mixin button {
							appearance: none;
						}
						@define-mixin large-button {
							@mixin button;
							font-size: 20px;
							div { color: pink; }
						}
						header {
							@mixin large-button;
						}
					`).To(BeParsedTo(`
						header {
							appearance: none;
							font-size: 20px;
							div { color: pink; }
						}
					`, "/foo.css"))
				})
			})

			Describe("url", func() {
				It("undefined mixin is passed through", func() {
					Expect(`
						header {
							@mixin unknown from url('/config/button.mixin.css');
						}
					`).To(BeParsedTo(`
						header {
							@mixin unknown from url('/config/button.mixin.css');
						}
					`, "/foo.css"))
				})

				It("mixin is replaced with defined mixin", Pending, func() {
					Expect(`
						header {
							@mixin large-button from url('/config/button.mixin.css');
						}
					`).To(BeParsedTo(``, "/foo.css"))
				})
			})
		})

		Describe("css modules", func() {
			It("path is not a css module", func() {
				Expect(`
					.title { color: green; }
				`).To(BeParsedTo(`
					.title { color: green; }
				`, "/foo.css"))
			})

			It("should rename classes", func() {
				Expect(`
					.title { color: green; }
				`).To(BeParsedTo(`
					.title43c30152 { color: green; }
				`))
			})

			It("should rename nested classes", func() {
				Expect(`
					.title {
						color: green;
						.subtitle { color: blue; }
					}
				`).To(BeParsedTo(`
					.title43c30152 {
						color: green;
						.subtitle43c30152 { color: blue; }
					}
				`))
			})

			It("should rename compound classes", func() {
				Expect(`
					.title.subtitle { color: green; }
				`).To(BeParsedTo(`
					.title43c30152.subtitle43c30152 { color: green; }
				`))
			})

			Describe("global function", func() {
				It("should not rename argument", func() {
					Expect(`
						.title { color: blue; }
						:global(.subtitle) { color: green; }
						.author { color: red; }
					`).To(BeParsedTo(`
						.title43c30152 { color: blue; }
						.subtitle { color: green; }
						.author43c30152 {	color: red; }
					`))
				})

				It("nested globals are ignored", Pending, func() {})

				It("nested classes should be renamed", func() {
					Expect(`
						:global(.subtitle) {
							color: green;
							.foo { color: orange; }
						}
					`).To(BeParsedTo(`
						.subtitle {
							color: green;
							.foo43c30152 { color: orange; }
						}
					`))
				})
			})

			Describe("global shorthand with argument", func() {
				It("should not rename argument", func() {
					Expect(`
						.title { color: blue; }
						:global
							.subtitle { color: green; }
						.author { color: red; }
					`).To(BeParsedTo(`
						.title43c30152 { color: blue; }
						.subtitle { color: green; }
						.author43c30152 { color: red; }
					`))
				})

				It("nested globals are ignored", Pending, func() {})

				It("nested classes should be renamed", func() {
					Expect(`
						:global .subtitle {
							color: green;
							.foo { color: orange; }
						}
					`).To(BeParsedTo(`
						.subtitle {
        			color: green;
        			.foo43c30152 { color: orange; }
						}
					`))
				})
			})

			Describe("global shorthand without argument", func() {
				It("should rename all children", func() {
					Expect(`
						:global {
							.subtitle {
								color: green;
								.foo { color: orange; }
							}
							.bar { color: blue; }
						}
						.author { color: red; }
					`).To(BeParsedTo(`
						.subtitle {
							color: green;
							.foo { color: orange; }
						}
						.bar { color: blue; }
						.author43c30152 { color: red; }
					`))
				})

				It("nested globals are ignored", Pending, func() {})
			})
		})
	})
})
