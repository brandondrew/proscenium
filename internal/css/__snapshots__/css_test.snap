
[TestParseCss/css_modules/path_is_not_a_css_module - 1]
.title {
    color: green;
}

---

[TestParseCss/css_modules/should_rename_classes - 1]
.title43c30152 {
    color: green;
}

---

[TestParseCss/css_modules/should_rename_nested_classes - 1]
.title43c30152 {
    color: green;
    .subtitle43c30152 {
        color: blue;
    }
}

---

[TestParseCss/css_modules/should_rename_compound_classes - 1]
.title43c30152.subtitle43c30152 {
    color: green;
}

---

[TestParseCss/css_modules/global_function/should_not_rename_function_argument - 1]
.title43c30152 {
    color: blue;
}
.subtitle {
    color: green;
}
.author43c30152 {
    color: red;
}

---

[TestParseCss/css_modules/global_function/nesting_in_function_body_should_be_renamed - 1]
.subtitle {
    color: green;
    .foo43c30152 {
        color: orange;
    }
}

---

[TestParseCss/css_modules/global_function/shorthand_with_argument/body_should_be_renamed - 1]
 .subtitle {
    color: green;
    .foo {
        color: orange;
    
}

---

[TestParseCss/css_modules/global_function/should_not_rename_argument - 1]
.title43c30152 {
    color: blue;
}
.subtitle {
    color: green;
}
.author43c30152 {
    color: red;
}

---

[TestParseCss/css_modules/global_function/nested_classes_should_be_renamed - 1]
.subtitle {
    color: green;
    .foo43c30152 {
        color: orange;
    }
}

---

[TestParseCss/css_modules/global_shorthand_with_argument/should_not_rename_argument - 1]
.title43c30152 {
    color: blue;
}
 .subtitle 
    color: green;
}
.author43c30152 {
    color: red;
}

---

[TestParseCss/css_modules/global_shorthand_with_argument/nested_classes_should_be_renamed - 1]
 .subtitle 
    color: green;
    .foo43c30152 {
        color: orange;
    }
}

---

[TestParseCss/css_modules/global_shorthand_without_argument/should_rename_all_children - 1]
 
    .subtitle {
        color: green;

        .foo {
            color: orange;
        }
    }
    .bar {
        color: blue;
    }
}
.author43c30152 {
    color: red;
}

---

[TestParseCss/should_pass_through_regular_css - 1]
body{}
---

[TestParseCss/mixins/undefined_mixin_is_passed_through - 1]
header {
    @mixin foo;
}

---

[TestParseCss/mixins/mixin_is_replaced_with_defined_mixin - 1]
header {
    link
}

---

[TestParseCss/mixins/local/undefined_mixin_is_passed_through - 1]
header {
    @mixin foo;
}

---

[TestParseCss/mixins/local/mixin_is_replaced_with_defined_mixin - 1]

header {
    font-size: 20px;
    div { color: pink; }
}

---

[TestParseCss/mixins/url/undefined_mixin_is_passed_through - 1]
header {
    @mixin unknown from url('/config/button.mixin.css');
}

---

[TestParseCss/mixins/url/mixin_is_replaced_with_defined_mixin - 1]
header {
    @mixin large-button;
}

---

[TestParseCss/mixins/local/mixin_must_be_defined_at_root_level - 1]
header {
    @define-mixin large-button {
        color: red;
    }
    div {
        @mixin foo;
    }
}

---

[TestParseCss/mixins/local/mixin_not_defined_at_root_level_is_passed_through - 1]
header {
    @define-mixin large-button {
        color: red;
    }
    div {
        @mixin foo;
    }
}

---

[TestParseCss/mixins/local/dependencies_are_fully_parsed - 1]

header {
    font-size: 20px;
    div { color: pink; }
}

---
